packages/panzer/disc-fe/src/Panzer_Workset.cpp

```diff
@@ -144,6 +143,73 @@ void WorksetDetails::setupNeeds(Teuchos::RCP<const shards::CellTopology> cell_to
+      const int faces_per_cell = cell_topology->getSubcellCount(space_dim-1);
+      const int points          = iv->surface_normals.extent_int(1);
+      const int points_per_face = points / faces_per_cell;
+      for (int virtual_cell_ordinal=0; virtual_cell_ordinal<_num_virtual_cells; virtual_cell_ordinal++)
+      {
+        const panzer::LocalOrdinal virtual_cell = virtual_cell_ordinal+num_real_cells;
+        const int face_ordinal = _face_connectivity->subcellForCell(virtual_cell, 0);
+        if (face_ordinal >= 0)
+        {
+          const int first_cell_for_face = _face_connectivity->cellForSubcell(face_ordinal, 0);
+          const panzer::LocalOrdinal other_side = (first_cell_for_face == virtual_cell) ? 1 : 0;
+          const panzer::LocalOrdinal real_cell = _face_connectivity->cellForSubcell(face_ordinal,other_side);
+          const panzer::LocalOrdinal face_in_real_cell = _face_connectivity->localSubcellForSubcell(face_ordinal,other_side);
+          TEUCHOS_ASSERT(real_cell < num_real_cells);
+          for (int point_ordinal=0; point_ordinal<points_per_face; point_ordinal++)
+          {
+            int virtual_cell_point = point_ordinal;
+            int real_cell_point = points_per_face * face_in_real_cell;
+            // the following arrays will be used to produce/store the rotation matrix below
+            double normal[3], transverse[3], binormal[3];
+            for(int i=0;i<3;i++)
+            {
+              normal[i]=0.;
+              transverse[i]=0.;
+              binormal[i]=0.;
+            }
+            
+            for (int d=0; d<space_dim; d++)
+            {
+              const auto n_d = iv->surface_normals(real_cell,real_cell_point,d);
+              iv->surface_normals(virtual_cell,virtual_cell_point,d) = -n_d;
+              normal[d] = -n_d;
+            }
+            
+            panzer::IntegrationValues2<double>::convertNormalToRotationMatrix(normal,transverse,binormal);
+            
+            for(int dim=0; dim<3; ++dim){
+              iv->surface_rotation_matrices(virtual_cell,virtual_cell_point,0,dim) = normal[dim];
+              iv->surface_rotation_matrices(virtual_cell,virtual_cell_point,1,dim) = transverse[dim];
+              iv->surface_rotation_matrices(virtual_cell,virtual_cell_point,2,dim) = binormal[dim];
+            }
+          }
+          // clear the other normals and rotation matrices for the virtual cell:
+          for (int point_ordinal=points_per_face; point_ordinal<points; point_ordinal++)
+          {
+            for (int dim=0; dim<space_dim; dim++)
+            {
+              iv->surface_normals(virtual_cell,point_ordinal,dim) = 0.0;
+            }
+            for(int dim1=0; dim1<3; ++dim1)
+            {
+              for(int dim2=0; dim2<3; ++dim2)
+              {
+                iv->surface_rotation_matrices(virtual_cell,point_ordinal,dim1,dim2) = 0;
+              }
+            }
+          }
+        }
+      }
+    }
```

Comment:
Could we move this chunk of code to a separate call? I have a branch that rewrites all of the workset IntegrationValues/BasisValues initializations and it would be nice to make this a modular interface outside of the workset.

Reply:
I have no objection to putting this code in a separate method at some point, but I don't fully understand your use case.  Wouldn't it make more sense for you to move this in your branch, after this is merged, since you know what you want here?  I'd like to get this merged in as soon as possible, since I have an EMPIRE MR that's blocked by this.